### PR TITLE
WIP: Redshift dependent subsample of lightcone/usmesh

### PR DESCRIPTION
**Pull request by rainwoodman**
Based on the initial C implementation of the sampling rate subroutines by YIci Zhong.

### CODE REVIEW

**Comment by yicizhong98** [comment]
I think the reason why test-lightcone fails can be mainly attributed to ell_lim we set. ell_lim=1000, 2000 give some failed results; but ell_lim=5000, 10000 pass all of the tests. The relative position between subsampling rate curve and subsamp_rate=1 shall somewhat be the origin for this problem.

File: libfastpm/cosmology.c [hunk 1]
```diff
@@ -480,6 +480,19 @@ double ComovingDistance(double a, FastPMCosmology * c) {
     return result;
 }
 
+/* Computes particle volume number density [1 / (Mpc/h)^3] 
+ * to reach the ell_lim resolution at given redshift */
+double VolumeDensityFromEll(double ell_lim, double z, FastPMCosmology * c)
+{
+    double theta_lim = M_PI / ell_lim;
+    double scale_fac = 1 / (1 + z);
+    double r = ComovingDistance(scale_fac,c) * HubbleDistance;
```
Comment (yicizhong98) [inline libfastpm/cosmology.c:489]: dh_factor

File: src/fastpm.c [hunk 2]
```diff
@@ -999,7 +999,17 @@ usmesh_ready_handler(FastPMUSMesh * mesh, FastPMLCEvent * lcevent, struct usmesh
     /* subsample */
     FastPMParticleMaskType * mask = fastpm_memory_alloc(lcevent->p->mem,
         "SubsampleMask", lcevent->p->np * sizeof(mask[0]), FASTPM_MEMORY_FLOATING);
-    fastpm_store_fill_subsample_mask(lcevent->p, CONF(prr->lua, particle_fraction), mask);
+    double particle_fraction = CONF(prr->lua, particle_fraction);
+    double a_mean = 0.5 * (lcevent->af + lcevent->ai);
```
Comment (yicizhong98) [inline src/fastpm.c:1003]: probably a bug

File: libfastpm/cosmology.c [hunk 3]
```diff
@@ -480,6 +503,21 @@ double ComovingDistance(double a, FastPMCosmology * c) {
     return result;
 }
 
+double ComovingDistance(double a, FastPMCosmology * c) {
+    struct FastPMComovingDistanceInterp * CDInterp = c->CDInterp;
+    double y;
+    int status;
+    status = gsl_interp_eval_e(CDInterp->interp, CDInterp->x, CDInterp->y, a, CDInterp->accel, &y);
```
Comment (yicizhong98) [inline libfastpm/cosmology.c:510]: Got error "srun: error: nidxxxx: task 0: Segmentation fault; srun: launch/slurm: _step_signal: Terminating StepId=xxxxxxx.x". I use gdb core mode to debug and got following message:
>>Core was generated by `/fastpm/tests/./testsubsample'.
>>Program terminated with signal SIGSEGV, Segmentation fault.
>>#0  gsl_interp_eval_e (interp=0x0, xa=0x7fffffff1ab8, ya=0x7fffffff3ac0, x=0.20000000000000001, a=0x0, y=0x7fffffff1998)
    at interp.c:131
131     interp.c: No such file or directory.

>>(gdb) bt
>>#0  gsl_interp_eval_e (interp=0x0, xa=0x7fffffff1ab8, ya=0x7fffffff3ac0, x=0.20000000000000001, a=0x0, y=0x7fffffff1998)
    at interp.c:131
>>#1  0x000000002000b2da in ComovingDistance (a=a@entry=0.20000000000000001, c=c@entry=0x7fffffff1a20) at cosmology.c:510
>>#2  0x00000000200018de in main (argc=<optimized out>, argv=<optimized out>) at testsubsample.c:64

seems like line 507-510 causes some storage problem.

File: src/fastpm.c (hunk 2 above)
Comment (rainwoodman) [reply src/fastpm.c]: Done. Using a_mean was problematic and we switched to use a per particle fraction.

File: libfastpm/cosmology.c (hunk 3 above)
Comment (rainwoodman) [reply libfastpm/cosmology.c]: We were missing a fastpm_cosmology_init in the tests.

**Pull request merged**